sqlx-sqlite/src/connection/mod.rs

```diff
@@ -112,6 +115,50 @@ impl SqliteConnection {
 
         Ok(LockedSqliteHandle { guard })
     }
+
+    /// Serializes the SQLite database to an in-memory buffer.
+    ///
+    /// This function allows you to serialize an SQLite database to an in-memory buffer.
+    /// The output is the same bytes as if the database were to be written to disk.
+    pub async fn serialize_to_buffer(&mut self) -> Result<SerializedDatabase, Error> {
+        let locked = self.lock_handle().await?;
+
+        unsafe {
+            let mut size: sqlite3_int64 = 0;
+            let data_ptr = sqlite3_serialize(
+                locked.guard.handle.as_ptr(),
+                std::ptr::null(),
+                &mut size as *mut _,
+                0,
+            );
+
+            Ok(SerializedDatabase::new(data_ptr, size as usize))
+        }
+    }
+    
+    /// Deserializes a byte slice into a read-only SQLite database.
+    ///
+    /// This function allows you to take a byte slice representing a serialized database,
+    /// and load it as a read-only SQLite database.
+    pub async fn deserialize_to_readonly_db(&mut self, data: &[u8]) -> Result<(), Error> {
```

Comment:
`sqlite3_deserialize()` taking an existing database connection and changing it to point at the deserialized database is an idiosyncrasy that we don't necessarily need to foist on the user.

IMO this should create a new `sqlite3` instance internally and return `Result<SqliteConnection, Error>`.

In that case it shouldn't be an `async fn`, though I'm not sure how much work is done in `sqlite3_deserialize()`--it may be wise to move that into a `spawn_blocking()` call, or just do it in the worker thread.

Reply:
On digging in to the `sqlite3_deserialize()` function, its implementation actually looks pretty trivial and so does not need any special handling. It might if it had to clean up a bunch of stuff on an existing connection, but if we're starting with a fresh one it should finish pretty much instantaneously.